tensorflow/core/grappler/optimizers/arithmetic_optimizer.cc

```diff
@@ -1783,6 +1808,133 @@ class HoistCWiseUnaryChainsStage : public ArithmeticOptimizerStage {
     return optimized_nodes_.find(node.name()) != optimized_nodes_.end();
   }
 
+  Status UpdateReshapesAndTails(const NodeDef* wrapped_node, const int port,
+                                PortNodeMap* reshapes, ChainLinkSet* tails,
+                                std::set<string>* ctrl_inputs) const {
+    NodeDef* unwrapped_node = nullptr;
+    TF_RETURN_IF_ERROR(UnwrapReshape(
+        wrapped_node, port, reshapes, ctrl_inputs, &unwrapped_node));
+    tails->insert(ChainLink(unwrapped_node, port));
+    return Status::OK();
+  }
+
+  Status UnwrapReshape(const NodeDef* wrapped_node, const int port,
+                       PortNodeMap* reshapes, std::set<string>* ctrl_inputs,
+                       NodeDef** return_node) const {
+    // If wrapped_node is not reshape, return the wrapped_node directly.
+    if (!IsReshape(*wrapped_node)) {
+      *return_node = const_cast<NodeDef*>(wrapped_node);
+      return Status::OK();
+    }
+
+    // Unwarp the all the reshape node.
+    if (node_is_concat_) {
+      TF_RETURN_IF_ERROR(UnwrapReshapeInput(wrapped_node, port, reshapes,
+                                            ctrl_inputs, return_node));
+    } else {
+      TF_RETURN_IF_ERROR(UnwrapReshapeOutput(wrapped_node, port, reshapes,
+                                             ctrl_inputs, return_node));
+    }
+    return Status::OK();
+  }
+
+  Status UnwrapReshapeInput(const NodeDef* wrapped_node, const int port,
+      PortNodeMap* reshapes, std::set<string>* ctrl_inputs,
+      NodeDef** return_node) const {
+    CHECK(IsReshape(*wrapped_node));
+
+    NodeDef* unwrapped_node = const_cast<NodeDef*>(wrapped_node);
+    // Record the last reshape node.
+    reshapes->insert(std::make_pair(port, unwrapped_node));
+
+    // Skip the reshape chain.
+    while (IsReshape(*unwrapped_node)) {
+      GatherControlInputs(unwrapped_node, ctrl_inputs);
+      TF_RETURN_IF_ERROR(
+          GetInputNode(unwrapped_node->input(0), &unwrapped_node));
+    }
+
+    *return_node = unwrapped_node;
+    return Status::OK();
+  }
+
+  Status UnwrapReshapeOutput(const NodeDef* wrapped_node, const int port,
+      PortNodeMap* reshapes, std::set<string>* ctrl_inputs,
+      NodeDef** return_node) const {
+    CHECK(IsReshape(*wrapped_node));
```

Comment:
`New CHECK or LOG(FATAL) statements inserted in code! Prefer using DCHECK, LOG(DFATAL), or DLOG(FATAL) instead`